### PR TITLE
Implemented a tile fetch queue.

### PR DESCRIPTION
**Pull request by manthey**
This is actually a general deferred object fetch queue.  It expects that no more than some number of Deferred objects should be processed at once.  For images in modern browsers, there is standard limit of six images per domain at a time.  This is the default for the fetch queue.  If more items are requested, they are further deferred and are loaded in a priority order of last-requested-first.  Items can be cancelled as well if they are no longer needed.

Also fixed the metrics that determine the load order of tiles.  There was an issue that tiles were passed and the function didn't know how to access tile.index.  This metric has been extended to have a concept of a 'bottomLevel', where anything lower than that level is lower priority.  This allows highly zoomed areas to show useful tiles first, rather than showing the absolute lowest tiles first.

This resolves issue #472.

### CODE REVIEW

**Comment by aashish24** [comment]
@manthey this is great addition :boom: I am doing the review now. Did you notice any performance improvement?

**Comment by manthey** [comment]
Yes, especially when moving the map rapidly.  The tiles you are looking at appear before the tiles that are hidden.

**Comment by jbeezley** [comment]
The fetch seems to work well, but I'm seeing issues with the tile layer.  It is trying to load invalid tiles now, for example:  `http://otile1.mqcdn.com/tiles/1.0.0/map/4/-2/9.png`.  The same issue occurs with and without tile wrapping.  It is a bit concerning that none of the tile layer tests detect this.  We should probably add some tests that cover this scenario.

A probably unrelated issue that this exposes is that CORS errors do not seem to be triggering error handlers on the tile promises.  When using tiles served locally, I just get a normal 404 and everything behaves as I would expect, but when getting tiles from a remote CDN.  The invalid urls raise CORS errors (at least on Chrome 47.0.2526.106) in the console

```
Image from origin 'http://otile1.mqcdn.com' has been blocked from loading by Cross-Origin Resource Sharing policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8082' is therefore not allowed access. The response had HTTP status code 404.
otile1.mqcdn.com/tiles/1.0.0/map/0/-1/0.png:1 GET http://otile1.mqcdn.com/tiles/1.0.0/map/0/-1/0.png 
```

When using the GL renderer an exception is raised when trying to load the texture [here](https://github.com/OpenGeoscience/vgl/blob/master/src/texture.js#L136).  After that the map will not redraw, but tile fetches continue to show up in the network view.

In Firefox (43.0.2), the 404's coming from the CDN are not triggering CORS errors.

**Comment by aashish24** [comment]
> http://otile1.mqcdn.com/tiles/1.0.0/map/4/-2/9.png

Just wanted to understand. There should not be a negative. How did we end up with a  negative index?

**Comment by aashish24** [comment]
@manthey I am going line by line so please allow me till this afternoon to review the branch

**Comment by manthey** [comment]
I had a typo.  In order to load tiles in the desired order, it is better to sort them before creating them, so that they get created in the order we want to load them.  When we wrap, we have the screen index (which can be negative) and the source index (which can't).  I had a typo which made the source unavailable.

File: src/core/fetchQueue.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+   *    needed.
+   */
+  //////////////////////////////////////////////////////////////////////////////
+  geo.fetchQueue = function (options) {
+    if (!(this instanceof geo.fetchQueue)) {
+      return new geo.fetchQueue(options);
+    }
+    options = options || {};
+    this._size = options.size || 6;
+    this._track = options.track || 600;
```
Comment (aashish24) [inline src/core/fetchQueue.js]: @manthey curious why did you pick up 600?

Comment (manthey) [inline src/core/fetchQueue.js]: As the comments say, picking a high number wastes memory and picking a low number wastes cpu.  The number should match the size of the tile cache, which we default to either 600 or 200 depending on whether we keep lower tiles or not (see comments in the tileLayer for why).  As we default to keeping lower tiles, that means the most common default will be 600.

Comment (aashish24) [inline src/core/fetchQueue.js]: thanks @manthey the code looks reasonable to me :+1: but I didn't run it.

**Comment by aashish24** [comment]
@manthey looks like we need to update the branch with master.

**Pull request merged**